vllm/sampling_params.py

```diff
@@ -33,6 +34,34 @@ class SamplingType(IntEnum):
 to sample from."""
 
 
+# maybe make msgspec?
+@dataclass
+class GuidedDecodingParams:
+    """One of these fields will be used to build a logit processor."""
+    json: Optional[Union[Dict, BaseModel, str]] = None
+    regex: Optional[str] = None
+    choice: Optional[List[str]] = None
+    grammar: Optional[str] = None
+    json_object: Optional[bool] = None
+
+    """These are other options that can be set"""
+    backend: Optional[str] = None
+    whitespace_pattern: Optional[str] = None
+
+    def __post_init__(self):
+        """Validate that some fields are mutually exclusive."""
+        guide_count = sum([
+            self.json is not None, self.regex is not None,
+            self.choice is not None, self.grammar is not None,
+            self.json_object is not None
+        ])
+        if guide_count > 1:
+            raise ValueError(
+                "You can only use one kind of guided decoding but multiple are "
+                f"specified: {self.__dict__}")
```

Comment:
Since only one of these fields is actually used, it may be better to make this a tagged union so that we don't have to allocate memory for the unused fields.

Reply:
Ah interesting, got any pointers on tagged unions in python?
A quick google showed this package which doesn't look production friendly: https://pypi.org/project/tagged-union/#history and I'm assuming that msgspec won't work for this case either as `str` is ambiguous for this struct: https://jcristharif.com/msgspec/supported-types.html#union-optional

We'd also then have to split out the `backend` and  `whitespace_pattern` fields which aren't mutually exclusive with the others, but that could be worth it if there's

Reply:
From my reading of tagged unions I think we'd need to implement something like this? Where each type has it's own class that specifies a tag, so that we can know which one to deserialize based on the tag

```
class Json(msgspec.Struct, tag=True):
    json: Union[Dict, BaseModel, str]
    
class Regex(msgspec.Struct, tag=True):
    regex: str

... (etc.)

class SamplingParams:
    ...
    guided_decoding: Union[Json, Regex, Choice, Grammar, JsonObject]
```

I could be way off though- let me know if I'm overcomplicating it and there's a much simpler way. As is I kinda prefer the easy readability of the `@dataclass`

Reply:
Yeah, from my understanding that should be how it's done in `msgspec`. It would be great if you can add some tests to check that the user is prevented from using multiple fields at once.

Reply:
Based on the performance though, I guess there's no need to use `msgspec` for now...

Reply:
added 👍